### PR TITLE
Modernize packaging

### PR DESCRIPTION
**Pull request by Exirel**
In preparation for Sopel 8 and with the ongoing effort to modernize packaging everywhere:

* Added a `pyproject.toml` for modern packaging
* Added a `Makefile` to perform `make develop build` ("release" could be added, if asked)
* Removed setuptools packaging information from `setup.cfg`
* Removed `setup.py`
* Moved code from `__init__.py` into `plugin.py`, as to prepare future change (with config and stuff)
* Added flake8, pylint, and pyroma linters
* Added a "if sopel 8" bit to properly manage IRC identifiers once Sopel 8 is out

To release on pypi, use `python -m twine upload dist/*` after using the `make build` command.

You can now use `make develop` to automatically update your virtualenv and have an editable install.

### CODE REVIEW

File: pyproject.toml
Comment (dgw) [inline pyproject.toml]: Is this left over from a copy-paste? As far as I know, you have not written any code for this plugin yet. 😝

I think it would make the most sense to credit the original author as acknowledged in the README and plugin file header—probably just as "Michael Yanovich", with no email address, so no one bothers them.

_Maybe_ I've done enough [in the past](https://github.com/sopel-irc/sopel-extras/commits/master/slap.py) to be considered a co-author, too. But I also wouldn't mind being only a maintainer.

File: pyproject.toml
Comment (dgw) [inline pyproject.toml:43]: Where's 3.11-3.12? It should work on those, assuming it's installed under a compatible Sopel version.

File: Makefile
Comment (dgw) [inline Makefile:4]: This is a standalone tool, no? It's not in dev-requirements.

File: dev-requirements.txt
Comment (dgw) [inline dev-requirements.txt]: I haven't run these checks or thought about customizing which warnings are enabled—but regardless of the settings in `setup.cfg`, doesn't having two linters seem redundant?

**Review by dgw — COMMENTED** [review]
The `Makefile` and added dev-tools is a bit more than just "Modernize packaging", but I know you mean well 🤪

File: pyproject.toml
Comment (Exirel) [reply pyproject.toml]: Clearly a copy/paste. Next PR, I'll add my name. 😄 

I think you've done enough regarding copyright.

File: pyproject.toml
Comment (Exirel) [reply pyproject.toml:43]: Untested so far. I've the same issue with other plugins, technically they should work fine with those Python versions, but since Sopel doesn't yet...

File: Makefile
Comment (Exirel) [reply Makefile:4]: Hm missed that one.

File: dev-requirements.txt
Comment (Exirel) [reply dev-requirements.txt]: Ah! Flake8 is a "if it doesn't pass, fix it", whereas pylint is more "here are some conventions you might want to check". For instance, pylint can give you a list of TODOs left in the code, or tell you about "hey, this could be a f-string", and it can be configured to ignore more or less these rules. At the end, it gives you a score you can look at to know where you stand.

And I do tend to have a perfect score on my projects... 😁 

Note that I didn't add any CI here as to not push that too far.

Comment (dgw) [reply dev-requirements.txt]: Ah, there's a nuance to it! Maybe we'll reach the point of adding `pylint` upstream in Sopel itself one day. There are undoubtedly code smells it would point out that aren't covered by our existing checks like CodeQL. But this is an entirely separate discussion!

> or tell you about "hey, this could be a f-string"

Bet I can guess the first rule you always configure `pylint` to ignore… 😁

**Comment by dgw** [comment]
@Exirel If you can get back to this soon-ish, that'd be helpful. Otherwise I will most likely implement my own subset of it—just the packaging and "if Sopel 8" parts, no extra tooling—for a version-bump compatible with Sopel 8.0.

File: pyproject.toml
Comment (Exirel) [reply pyproject.toml:43]: I've added them since Sopel 8 is out!

File: Makefile
Comment (Exirel) [reply Makefile:4]: Fixed.

File: dev-requirements.txt
Comment (Exirel) [reply dev-requirements.txt]: I'll see to add tests & CI in a future PR. :)

**Comment by Exirel** [comment]
> @Exirel If you can get back to this soon-ish, that'd be helpful.

Yup, done!

_Edit_: I'm waiting on your approval to squash these commits.

**Review by dgw — APPROVED** [review]
Have I tested all the extra build/dev tools? No, lol. But I trust you to create a better dev environment than me, at this point.

About all I can say about this rework is that it's probably not necessary to keep the py2 stuff (e.g. encoding comment) around. sopel-slap 0.1 will work with Sopel 7.x indefinitely, and 0.2 could just be for 8+ only now that 8.0 has released.

File: pyproject.toml
Comment (dgw) [inline pyproject.toml]: Oops, forgot to call this out. 3.7 is EOL so let's not bother advertising support.

**Comment by dgw** [comment]
While awaiting a squash, I have set up Trusted Publishing for this repository so I can easily cut a new release post-merge. 😺

**Comment by Exirel** [comment]
Github is annoying, it didn't send me the approval email notification, so I didn't see it until now. Will see if I can manage that tonight.

**Comment by Exirel** [comment]
Squashed and I removed 3.7 as well.

File: pyproject.toml
Comment (dgw) [inline pyproject.toml:50]: I thought I'd suggested leaving only Sopel 8 support, but maybe not. Regardless, since this plugin imports `sopel.plugin` it must require 7.1+ at minimum.


```suggestion
    "sopel>=7.1",
```

**Review by dgw — COMMENTED** [review]
Little backward-compatibility nit, sorry 😅

File: pyproject.toml
Comment (dgw) [reply pyproject.toml:50]: Hmm, to heck with this; the mistake is a holdover from when I initially put the plugin together. I'll fix it during release.

**Pull request merged**
